### PR TITLE
Add rocksdb

### PR DESCRIPTION
**Pull request by Stumble**
https://github.com/tendermint/tendermint/issues/3831

Using my fork of gorocksdb from https://github.com/tecbot/gorocksdb for two reasons: 1. The upstream repo encourages vendoring for API stability. 2. The default LDFLAGS setting in upstream repo enforces linking unnecessary compression libraries(https://github.com/tecbot/gorocksdb/blob/162552197222a834c6857385ae890e4decdd5049/dynflag.go#L5), so that if the library is not installed, link will fail. IMHO, tendermint should fork this wrapper and use it here.

To run tests:
First, install rocksdb,
```
homedir="$PWD"
sudo apt-get update
sudo apt-get install -y build-essential gcc g++ make libgflags-dev libsnappy-dev
wget -O rocks.tar.gz https://github.com/facebook/rocksdb/archive/v6.1.2.tar.gz && \
  tar -zxvf rocks.tar.gz && \
  cd rocksdb-6.1.2/ && \
  make static_lib -j4 && \
  sudo cp librocksdb.a /usr/local/lib/ && \
  cd include/ && \
  sudo rm -rf /usr/local/include/rocksdb && \
  sudo cp -r rocksdb /usr/local/include/ && \
  sudo ldconfig
cd "$homedir"
rm rocks.tar.gz
rm -rf rocksdb-6.1.2
```
Run tests:
```
GO_LDFLAGS="-lsnappy" go test -v -tags "rocksdb"
```

### CODE REVIEW

File: db/db.go [hunk 1]
```diff
@@ -32,6 +32,11 @@ const (
 	//   - may be faster is some use-cases (random reads - indexer)
 	//   - use boltdb build tag (go build -tags boltdb)
 	BoltDBBackend DBBackendType = "boltdb"
+	// RocksDBBackend represents rocksdb (uses tecbot's wrapper)
+	//   - EXPERIMENTAL
+	//   - requires gcc
+	//   - use boltdb build tag (go build -tags rocksdb)
```
Comment (melekes) [inline db/db.go]:
```suggestion
	//   - use rocksdb build tag (go build -tags rocksdb)
```

File: db/c_level_db_test.go [hunk 2]
```diff
@@ -10,7 +10,7 @@ import (
 
 	"github.com/stretchr/testify/assert"
 
-	cmn "github.com/tendermint/tm-cmn/libs/common"
+	cmn "github.com/tendermint/tm-cmn/common"
```
Comment (melekes) [inline db/c_level_db_test.go]: this got changed I think. Sorry, we're in a process of separating some libraries.

File: db/rocks_db.go [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+func NewRocksDBWithOptions(name string, dir string, opts *gorocksdb.Options) (*RocksDB, error) {
+	dbPath := filepath.Join(dir, name+".db")
+	db, err := gorocksdb.OpenDb(opts, dbPath)
+	if err != nil {
+		return nil, err
+	}
+	ro := gorocksdb.NewDefaultReadOptions()
+	wo := gorocksdb.NewDefaultWriteOptions()
+	woSync := gorocksdb.NewDefaultWriteOptions()
+	woSync.SetSync(true)
```
Comment (melekes) [inline db/rocks_db.go]: what this line does?

File: db/rocks_db.go [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+	}
+	return database, nil
+}
+
+// Implements DB.
+func (db *RocksDB) Get(key []byte) []byte {
+	key = nonNilBytes(key)
+	res, err := db.db.Get(db.ro, key)
+	if err != nil {
+		res.Free()
```
Comment (melekes) [inline db/rocks_db.go]: why not put this above with `defer`? that way we'll avoid duplication s.Free() in moveSliceToBytes and here.

Comment (Stumble) [reply db/rocks_db.go]: turns out `res.Free()` when `err != nil` is not necessary, removed. I would prefer to have a function that moves c bytes to go slice.

File: db/rocks_db.go (hunk 3 above)
Comment (Stumble) [reply db/rocks_db.go]: Do you mean these variables? read/write options?

Comment (melekes) [reply db/rocks_db.go]: I mean `woSync.SetSync(true)`

**Comment by melekes** [comment]
> Using my fork of gorocksdb from https://github.com/tecbot/gorocksdb for two reasons:

Have you tried to contribute back to original lib?

File: db/rocks_db.go (hunk 3 above)
Comment (Stumble) [reply db/rocks_db.go]: `woSync.SetSync(true)` prepares a sync write option. https://godoc.org/github.com/tecbot/gorocksdb#WriteOptions.SetSync

**Comment by Stumble** [comment]
> > Using my fork of gorocksdb from https://github.com/tecbot/gorocksdb for two reasons:
> 
> Have you tried to contribute back to original lib?

Nope, as vendoring was suggested for gorocksdb anyway. I just sent an issue in the upstream, let's see how it goes.

**Comment by Stumble** [comment]
@melekes , the upstream has not respond to the issue yet. https://github.com/tecbot/gorocksdb/issues/167

**Comment by Stumble** [comment]
Rocksdb's stats does not work with https://github.com/cosmos/cosmos-sdk/blob/e4c8bd72b72b18035ede103cb4a47da5825d330e/server/export.go#L95.

File: db.go [hunk 5]
```diff
@@ -32,6 +32,11 @@ const (
 	//   - may be faster is some use-cases (random reads - indexer)
 	//   - use boltdb build tag (go build -tags boltdb)
 	BoltDBBackend DBBackendType = "boltdb"
+	// RocksDBBackend represents rocksdb (uses tecbot's wrapper)
```
Comment (melekes) [inline db.go]:
```suggestion
	// RocksDBBackend represents rocksdb (uses Stumble fork github.com/stumble/gorocksdb of the tecbot wrapper)
```

**Comment by marbar3778** [comment]
@Stumble it seems gorocksdb has picked up steam could you rebase your fork until your PR get merged?

**Comment by Stumble** [comment]
@marbar3778 The PR is merged, I suggest we switch to the original branch.

**Pull request merged**